Look up runtime libraries in SDK

beccadax opened the pull request:
In #23175, we started looking in the SDK for swiftmodules, but we want to look for the libraries there too while linking. Fixes <rdar://problem/52059706>.

~~(This PR is not quite fully baked, but I want to kick off Linux tests while I read over the code and check a few details about how this should behave.)~~

This PR ought to be ready to go, but I could use confirmation that it's actually implementing the desired behavior.

beccadax committed:
Look up runtime libraries in SDK

In #23175, we started looking in the SDK for swiftmodules, but we want to look for the dylibs there too. Fixes <rdar://problem/52059706>.

brentdax commented:
@swift-ci please smoke test

beccadax committed:
Clean up driver linker invocation code

Refactor to merge some very similar code in different branches. This ended up slightly changing the order of arguments in Linux linker commands.

beccadax committed:
Explicitly set SDK in interpreter test

Gives us consistent results between macOS and Linux.

brentdax commented:
@swift-ci please test

brentdax commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -444,23 +442,34 @@ toolchains::Darwin::constructInvocation(const DynamicLinkJobAction &job,
+
+  SmallVector<std::string, 4> RuntimeLibPaths;
+  getRuntimeLibraryPaths(RuntimeLibPaths, context.Args,
+                         context.OI.SDKPath, /*Shared=*/!wantsStaticStdlib);
```
Doing this properly will require a little more work than I want to do right now. I've filed https://bugs.swift.org/browse/SR-11015 to track the idea.

beccadax committed:
Undo Linux linker order changes

Turns out the order of the linker flags matters on Linux; see apple/swift#9958. I don’t know if the change I previously made would break something, but it’s not worth risking it.

brentdax commented:
@swift-ci please smoke test

compnerd commented:
@swift-ci please test Windows platform

shahmishal commented:
@swift-ci please test Windows platform

beccadax committed:
Fix a second Linux test

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please test Windows platform
